File: app/src/main/java/mx/clubsanfrancisco/golfgps/ScorecardImage.kt
package mx.clubsanfrancisco.golfgps

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Typeface
import androidx.core.content.FileProvider
import java.io.File
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Dibuja la tarjeta de la ronda actual en un PNG y arma un Intent para
 * compartirla (WhatsApp, correo, etc.). Todo con Canvas de Android: no depende
 * de la jerarquía de Compose, así que funciona aunque la vista no esté montada.
 */
object ScorecardImage {

    // Paleta de la imagen (independiente del tema de la app, para que se vea
    // igual y legible en WhatsApp con fondo claro).
    private const val GREEN = 0xFF1B5E3F.toInt()
    private const val GREEN_SOFT = 0xFFE6F0EA.toInt()
    private const val WHITE = 0xFFFFFFFF.toInt()
    private const val INK = 0xFF1A1A1A.toInt()
    private const val GRID = 0xFFD9D9D9.toInt()
    private const val MUTED = 0xFF6B6B6B.toInt()
    private const val MARK = 0xFF9BA8A0.toInt()      // aro/cuadro bajo/sobre par
    private const val STRIPE = 0xFFF3F8F5.toInt()    // franja de filas alternas

    fun shareIntent(context: Context, vm: GolfViewModel): Intent? {
        val players = vm.players.toList()
        if (players.isEmpty()) return null
        val bmp = render(vm, players)
        val dir = File(context.cacheDir, "shared").apply { mkdirs() }
        val file = File(dir, "tarjeta_sfgolf.png")
        FileOutputStream(file).use { bmp.compress(Bitmap.CompressFormat.PNG, 100, it) }
        val uri = FileProvider.getUriForFile(context, "${context.packageName}.fileprovider", file)
        return Intent(Intent.ACTION_SEND).apply {
            type = "image/png"
            putExtra(Intent.EXTRA_STREAM, uri)
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
    }

    private fun paint(size: Float, color: Int, bold: Boolean, center: Boolean = false) =
        Paint(Paint.ANTI_ALIAS_FLAG).apply {
            textSize = size
            this.color = color
            typeface = Typeface.create(Typeface.DEFAULT, if (bold) Typeface.BOLD else Typeface.NORMAL)
            if (center) textAlign = Paint.Align.CENTER
        }

    private const val W = 1080
    private const val PAD = 36f
    private const val LABEL_W = 190f
    private const val ROW_H = 62f
    /** Las columnas OUT / TOTAL son más anchas que las de hoyo (no se cortan). */
    private const val SUM_COL = 1.45f
    private const val FOOTER_H = 50f

    private fun render(vm: GolfViewModel, players: List<Player>): Bitmap {
        val n = players.size
        val titleH = 168f
        val blockH = ROW_H * (2 + n)          // fila Hoyo + Par + jugadores
        val gap = 28f
        val footHeadH = 54f
        val footH = footHeadH + ROW_H * n
        // gap tras el título + entre bloques + antes del resumen = 3 gaps.
        val height = (titleH + gap + blockH + gap + blockH + gap + footH + FOOTER_H + PAD).toInt()

        val bmp = Bitmap.createBitmap(W, height, Bitmap.Config.ARGB_8888)
        val c = Canvas(bmp)
        c.drawColor(WHITE)

        // ---- Encabezado ----
        c.drawRect(0f, 0f, W.toFloat(), titleH, paint(1f, GREEN, false).apply { style = Paint.Style.FILL })
        c.drawText("⛳ SF Golf GPS", PAD, 74f, paint(52f, WHITE, true))
        c.drawText(CourseData.CLUB_NAME + " · Par ${CourseData.totalPar}", PAD, 116f, paint(30f, WHITE, false))
        val fecha = SimpleDateFormat("EEEE d 'de' MMMM, yyyy · h:mm a", Locale("es", "MX")).format(Date())
        c.drawText(fecha.replaceFirstChar { it.uppercase() }, PAD, 150f, paint(26f, 0xFFCDE4D6.toInt(), false))

        var y = titleH + gap
        y = drawBlock(c, players, 0 until 9, "OUT", null, y, n)
        y += gap
        y = drawBlock(c, players, 9 until 18, "IN", "TOTAL", y, n)
        y += gap

        // ---- Resumen por jugador ----
        c.drawText("Resumen", PAD, y + 36f, paint(34f, GREEN, true))
        y += footHeadH
        players.forEach { p ->
            drawSummary(c, p, y)
            y += ROW_H
        }

        // ---- Pie ----
        c.drawText(
            "⛳ SF Golf GPS · ${CourseData.CLUB_NAME}",
            W / 2f, y + FOOTER_H - 8f,
            paint(22f, MUTED, false, center = true)
        )
        return bmp
    }

    /** Un bloque de 9 hoyos (ida o vuelta) con su fila Hoyo, Par y cada jugador. */
    private fun drawBlock(
        c: Canvas, players: List<Player>, range: IntRange,
        sumLabel: String, totLabel: String?, top: Float, n: Int
    ): Float {
        val extraCols = if (totLabel != null) 2 else 1  // OUT, o IN+TOTAL
        val units = 9 + extraCols * SUM_COL
        val cellW = (W - 2 * PAD - LABEL_W) / units
        val sumW = cellW * SUM_COL
        val x0 = PAD

        // Centros de columna: hoyos 0-8, luego OUT/IN y TOTAL (más anchas).
        fun holeX(i: Int) = x0 + LABEL_W + i * cellW + cellW / 2
        val sumX = x0 + LABEL_W + 9 * cellW + sumW / 2
        val totX = x0 + LABEL_W + 9 * cellW + sumW + sumW / 2

        // Fila de hoyos (fondo verde)
        var y = top
        c.drawRect(x0, y, W - PAD, y + ROW_H, paint(1f, GREEN, false).apply { style = Paint.Style.FILL })
        c.drawText("Hoyo", x0 + 14f, y + 40f, paint(30f, WHITE, true))
        range.forEachIndexed { i, h ->
            c.drawText("${h + 1}", holeX(i), y + 40f, paint(30f, WHITE, true, center = true))
        }
        c.drawText(sumLabel, sumX, y + 40f, paint(28f, WHITE, true, center = true))
        totLabel?.let { c.drawText(it, totX, y + 40f, paint(24f, WHITE, true, center = true)) }

        // Fila de par (fondo suave)
        y += ROW_H
        c.drawRect(x0, y, W - PAD, y + ROW_H, paint(1f, GREEN_SOFT, false).apply { style = Paint.Style.FILL })
        c.drawText("Par", x0 + 14f, y + 40f, paint(28f, MUTED, true))
        range.forEachIndexed { i, h ->
            c.drawText("${CourseData.holes[h].par}", holeX(i), y + 40f, paint(28f, MUTED, true, center = true))
        }
        val parSum = range.sumOf { CourseData.holes[it].par }
        c.drawText("$parSum", sumX, y + 40f, paint(28f, MUTED, true, center = true))
        totLabel?.let { c.drawText("${CourseData.totalPar}", totX, y + 40f, paint(28f, MUTED, true, center = true)) }

        // Filas por jugador (franjas alternas para leer mejor)
        players.forEachIndexed { pi, p ->
            y += ROW_H
            if (pi % 2 == 1) {
                c.drawRect(x0, y, W - PAD, y + ROW_H, paint(1f, STRIPE, false).apply { style = Paint.Style.FILL })
            }
            c.drawText(p.name.take(11), x0 + 14f, y + 40f, paint(28f, INK, true))
            range.forEachIndexed { i, h ->
                val s = p.strokes[h]
                val cx = holeX(i)
                if (s > 0) {
                    drawScoreMark(c, cx, y + ROW_H / 2, s - CourseData.holes[h].par)
                    c.drawText("$s", cx, y + 40f, paint(30f, INK, true, center = true))
                } else {
                    c.drawText("·", cx, y + 40f, paint(30f, GRID, true, center = true))
                }
            }
            val outSum = range.sumOf { p.strokes[it] }
            c.drawText(if (outSum > 0) "$outSum" else "·", sumX, y + 40f, paint(30f, INK, true, center = true))
            totLabel?.let {
                val tot = p.total()
                c.drawText(if (tot > 0) "$tot" else "·", totX, y + 40f, paint(32f, GREEN, true, center = true))
            }
        }

        // Borde del bloque
        c.drawRect(x0, top, W - PAD, y + ROW_H, paint(1f, GRID, false).apply {
            style = Paint.Style.STROKE; strokeWidth = 2f
        })
        return y + ROW_H
    }

    /** Círculo para bajo par, cuadro para sobre par (como en la app), en gris neutro. */
    private fun drawScoreMark(c: Canvas, cx: Float, cy: Float, diff: Int) {
        if (diff == 0) return
        val r = 24f
        val p = paint(1f, MARK, false).apply {
            style = Paint.Style.STROKE; strokeWidth = 2.5f
        }
        if (diff < 0) c.drawCircle(cx, cy, r, p)
        else c.drawRoundRect(RectF(cx - r, cy - r, cx + r, cy + r), 6f, 6f, p)
    }

    private fun drawSummary(c: Canvas, p: Player, top: Float) {
        val rel = p.relativeToPar()
        val relTxt = when {
            p.playedHoles() == 0 -> "–"
            rel == 0 -> "E"
            rel > 0 -> "+$rel"
            else -> "$rel"
        }
        val bits = buildList {
            add("${p.total()} ($relTxt)")
            val pts = p.stablefordPoints()
            if (pts > 0) add("$pts pts" + if (p.hcp > 0) " · hcp ${p.hcp}" else "")
            val (fh, fa) = p.firStats()
            if (fa > 0) add("FIR $fh/$fa")
            if (p.girTracked() > 0) add("GIR ${p.girCount()}/${p.girTracked()}")
            if (p.totalPutts() > 0) add("${p.totalPutts()} putts")
        }
        c.drawText(p.name.take(12), PAD, top + 40f, paint(30f, INK, true))
        // El texto se encoge hasta caber en el ancho disponible (no se corta).
        val text = bits.joinToString("  ·  ")
        val maxW = W - PAD - (PAD + 250f)
        val pnt = paint(26f, MUTED, false)
        while (pnt.measureText(text) > maxW && pnt.textSize > 17f) pnt.textSize -= 1f
        c.drawText(text, PAD + 250f, top + 40f, pnt)
    }
}

File: ios/SFGolfGPS/ScorecardImage.swift
import UIKit

/// Dibuja la tarjeta de la ronda actual en un PNG para compartirla
/// (WhatsApp, correo, etc.). Todo con Core Graphics: mismo layout y paleta
/// que la versión Android (ScorecardImage.kt) para que la imagen sea idéntica
/// en ambas plataformas.
enum ScorecardImage {

    // Paleta de la imagen (independiente del tema de la app, para que se vea
    // igual y legible en WhatsApp con fondo claro).
    private static let green = UIColor(red: 0x1B / 255, green: 0x5E / 255, blue: 0x3F / 255, alpha: 1)
    private static let greenSoft = UIColor(red: 0xE6 / 255, green: 0xF0 / 255, blue: 0xEA / 255, alpha: 1)
    private static let headerSub = UIColor(red: 0xCD / 255, green: 0xE4 / 255, blue: 0xD6 / 255, alpha: 1)
    private static let ink = UIColor(red: 0x1A / 255, green: 0x1A / 255, blue: 0x1A / 255, alpha: 1)
    private static let grid = UIColor(red: 0xD9 / 255, green: 0xD9 / 255, blue: 0xD9 / 255, alpha: 1)
    private static let muted = UIColor(red: 0x6B / 255, green: 0x6B / 255, blue: 0x6B / 255, alpha: 1)
    private static let mark = UIColor(red: 0x9B / 255, green: 0xA8 / 255, blue: 0xA0 / 255, alpha: 1)
    private static let stripe = UIColor(red: 0xF3 / 255, green: 0xF8 / 255, blue: 0xF5 / 255, alpha: 1)

    private static let W: CGFloat = 1080
    private static let pad: CGFloat = 36
    private static let labelW: CGFloat = 190
    private static let rowH: CGFloat = 62
    /// Las columnas OUT / TOTAL son más anchas que las de hoyo (no se cortan).
    private static let sumCol: CGFloat = 1.45
    private static let footerH: CGFloat = 50

    private static func attrs(_ size: CGFloat, _ color: UIColor, bold: Bool) -> [NSAttributedString.Key: Any] {
        [.font: UIFont.systemFont(ofSize: size, weight: bold ? .bold : .regular),
         .foregroundColor: color]
    }

    private static func drawText(_ text: String, x: CGFloat, y: CGFloat,
                                 size: CGFloat, color: UIColor, bold: Bool, center: Bool = false) {
        let a = attrs(size, color, bold: bold)
        let s = NSAttributedString(string: text, attributes: a)
        let w = s.size().width
        // y es la baseline (como en el Canvas de Android): compensa el ascent.
        let font = a[.font] as! UIFont
        let origin = CGPoint(x: center ? x - w / 2 : x, y: y - font.ascender)
        s.draw(at: origin)
    }

    static func render(model: GolfModel) -> UIImage? {
        let players = model.players
        if players.isEmpty { return nil }

        let n = CGFloat(players.count)
        let titleH: CGFloat = 168
        let blockH = rowH * (2 + n)          // fila Hoyo + Par + jugadores
        let gap: CGFloat = 28
        let footHeadH: CGFloat = 54
        let footH = footHeadH + rowH * n
        let height = titleH + gap + blockH + gap + blockH + gap + footH + footerH + pad

        let fmt = UIGraphicsImageRendererFormat()
        fmt.scale = 1
        let renderer = UIGraphicsImageRenderer(size: CGSize(width: W, height: height), format: fmt)
        return renderer.image { ctx in
            let c = ctx.cgContext
            c.setFillColor(UIColor.white.cgColor)
            c.fill(CGRect(x: 0, y: 0, width: W, height: height))

            // ---- Encabezado ----
            c.setFillColor(green.cgColor)
            c.fill(CGRect(x: 0, y: 0, width: W, height: titleH))
            drawText("⛳ SF Golf GPS", x: pad, y: 74, size: 52, color: .white, bold: true)
            drawText("\(CourseData.clubName) · Par \(CourseData.totalPar)", x: pad, y: 116, size: 30, color: .white, bold: false)
            let df = DateFormatter()
            df.locale = Locale(identifier: "es_MX")
            df.dateFormat = "EEEE d 'de' MMMM, yyyy · h:mm a"
            let fecha = df.string(from: Date())
            drawText(fecha.prefix(1).uppercased() + fecha.dropFirst(), x: pad, y: 150, size: 26, color: headerSub, bold: false)

            var y = titleH + gap
            y = drawBlock(c, players, range: 0..<9, sumLabel: "OUT", totLabel: nil, top: y)
            y += gap
            y = drawBlock(c, players, range: 9..<18, sumLabel: "IN", totLabel: "TOTAL", top: y)
            y += gap

            // ---- Resumen por jugador ----
            drawText("Resumen", x: pad, y: y + 36, size: 34, color: green, bold: true)
            y += footHeadH
            for p in players {
                drawSummary(p, top: y)
                y += rowH
            }

            // ---- Pie ----
            drawText("⛳ SF Golf GPS · \(CourseData.clubName)",
                     x: W / 2, y: y + footerH - 8, size: 22, color: muted, bold: false, center: true)
        }
    }

    /// Un bloque de 9 hoyos (ida o vuelta) con su fila Hoyo, Par y cada jugador.
    private static func drawBlock(_ c: CGContext, _ players: [Player], range: Range<Int>,
                                  sumLabel: String, totLabel: String?, top: CGFloat) -> CGFloat {
        let extraCols: CGFloat = totLabel != nil ? 2 : 1  // OUT, o IN+TOTAL
        let units = 9 + extraCols * sumCol
        let cellW = (W - 2 * pad - labelW) / units
        let sumW = cellW * sumCol
        let x0 = pad

        // Centros de columna: hoyos 0-8, luego OUT/IN y TOTAL (más anchas).
        func holeX(_ i: Int) -> CGFloat { x0 + labelW + CGFloat(i) * cellW + cellW / 2 }
        let sumX = x0 + labelW + 9 * cellW + sumW / 2
        let totX = x0 + labelW + 9 * cellW + sumW + sumW / 2

        // Fila de hoyos (fondo verde)
        var y = top
        c.setFillColor(green.cgColor)
        c.fill(CGRect(x: x0, y: y, width: W - 2 * pad, height: rowH))
        drawText("Hoyo", x: x0 + 14, y: y + 40, size: 30, color: .white, bold: true)
        for (i, h) in range.enumerated() {
            drawText("\(h + 1)", x: holeX(i), y: y + 40, size: 30, color: .white, bold: true, center: true)
        }
        drawText(sumLabel, x: sumX, y: y + 40, size: 28, color: .white, bold: true, center: true)
        if let totLabel {
            drawText(totLabel, x: totX, y: y + 40, size: 24, color: .white, bold: true, center: true)
        }

        // Fila de par (fondo suave)
        y += rowH
        c.setFillColor(greenSoft.cgColor)
        c.fill(CGRect(x: x0, y: y, width: W - 2 * pad, height: rowH))
        drawText("Par", x: x0 + 14, y: y + 40, size: 28, color: muted, bold: true)
        for (i, h) in range.enumerated() {
            drawText("\(CourseData.holes[h].par)", x: holeX(i), y: y + 40, size: 28, color: muted, bold: true, center: true)
        }
        let parSum = range.reduce(0) { $0 + CourseData.holes[$1].par }
        drawText("\(parSum)", x: sumX, y: y + 40, size: 28, color: muted, bold: true, center: true)
        if totLabel != nil {
            drawText("\(CourseData.totalPar)", x: totX, y: y + 40, size: 28, color: muted, bold: true, center: true)
        }

        // Filas por jugador (franjas alternas para leer mejor)
        for (pi, p) in players.enumerated() {
            y += rowH
            if pi % 2 == 1 {
                c.setFillColor(stripe.cgColor)
                c.fill(CGRect(x: x0, y: y, width: W - 2 * pad, height: rowH))
            }
            drawText(String(p.name.prefix(11)), x: x0 + 14, y: y + 40, size: 28, color: ink, bold: true)
            for (i, h) in range.enumerated() {
                let s = p.strokes[h]
                let cx = holeX(i)
                if s > 0 {
                    drawScoreMark(c, cx: cx, cy: y + rowH / 2, diff: s - CourseData.holes[h].par)
                    drawText("\(s)", x: cx, y: y + 40, size: 30, color: ink, bold: true, center: true)
                } else {
                    drawText("·", x: cx, y: y + 40, size: 30, color: grid, bold: true, center: true)
                }
            }
            let outSum = range.reduce(0) { $0 + p.strokes[$1] }
            drawText(outSum > 0 ? "\(outSum)" : "·", x: sumX, y: y + 40, size: 30, color: ink, bold: true, center: true)
            if totLabel != nil {
                let tot = p.total()
                drawText(tot > 0 ? "\(tot)" : "·", x: totX, y: y + 40, size: 32, color: green, bold: true, center: true)
            }
        }

        // Borde del bloque
        c.setStrokeColor(grid.cgColor)
        c.setLineWidth(2)
        c.stroke(CGRect(x: x0, y: top, width: W - 2 * pad, height: y + rowH - top))
        return y + rowH
    }

    /// Círculo para bajo par, cuadro para sobre par (como en la app), en gris neutro.
    private static func drawScoreMark(_ c: CGContext, cx: CGFloat, cy: CGFloat, diff: Int) {
        if diff == 0 { return }
        let r: CGFloat = 24
        c.setStrokeColor(mark.cgColor)
        c.setLineWidth(2.5)
        let rect = CGRect(x: cx - r, y: cy - r, width: r * 2, height: r * 2)
        if diff < 0 {
            c.strokeEllipse(in: rect)
        } else {
            c.addPath(UIBezierPath(roundedRect: rect, cornerRadius: 6).cgPath)
            c.strokePath()
        }
    }

    private static func drawSummary(_ p: Player, top: CGFloat) {
        let rel = p.relativeToPar()
        let relTxt: String = {
            if p.playedHoles() == 0 { return "–" }
            if rel == 0 { return "E" }
            return rel > 0 ? "+\(rel)" : "\(rel)"
        }()
        var bits = ["\(p.total()) (\(relTxt))"]
        let pts = p.stablefordPoints()
        if pts > 0 { bits.append("\(pts) pts" + (p.hcp > 0 ? " · hcp \(p.hcp)" : "")) }
        let (fh, fa) = p.firStats()
        if fa > 0 { bits.append("FIR \(fh)/\(fa)") }
        if p.girTracked() > 0 { bits.append("GIR \(p.girCount())/\(p.girTracked())") }
        if p.totalPutts() > 0 { bits.append("\(p.totalPutts()) putts") }

        drawText(String(p.name.prefix(12)), x: pad, y: top + 40, size: 30, color: ink, bold: true)
        // El texto se encoge hasta caber en el ancho disponible (no se corta).
        let text = bits.joined(separator: "  ·  ")
        let maxW = W - pad - (pad + 250)
        var size: CGFloat = 26
        while NSAttributedString(string: text, attributes: attrs(size, muted, bold: false)).size().width > maxW
                && size > 17 {
            size -= 1
        }
        drawText(text, x: pad + 250, y: top + 40, size: size, color: muted, bold: false)
    }
}
